### PR TITLE
Generalizing cross validation for other data types and Trainers.

### PR DESCRIPTION
**Pull request by crb3874**
Refs #21399

This PR builds on the recent PR for implementing cross-validation capabilities in SurrogateTrainers (https://github.com/idaholab/moose/pull/21405). The main purpose is generalizing for other `Trainers`, and other types of input/output data.

- Refactored `SurrogateTrainer` and several derived classes to handle Reporter gathering for common (`Real`, `std::vector<Real>`) response and predictor types in the base class. This removed some duplicated code in other Trainers in the module.
- Added some API functions and protected member variables in `SurrogateTrainer` for accessing common types of training data in derived Trainers.
- Refactored several Trainers to make them compatible with cross validation.
- Added `std::vector<Real>` response type support to `NearestPointTrainer` (this isn't particularly important, I implemented this for an earlier, now un-used, approach to handling the other points in this PR. I figured I'd leave it in, but it can easily be removed).
- Added documentation for the changes to the `SurrogateTrainer` syntax.
- Added tests for CV with all trainer types, and for CV with `std::vector<Real>` response types.

### CODE REVIEW

File: modules/stochastic_tools/test/src/reporters/CrossValidationScores.C [hunk 1]
```diff
@@ -18,24 +18,33 @@ CrossValidationScores::validParams()
...
 
 void
 CrossValidationScores::execute()
 {
+  unsigned int n_r = _response_indices.size();
   for (unsigned int m = 0; m < _models.size(); ++m)
-    (*_cv_scores[m]) = _models[m]->getModelData<std::vector<Real>>("cv_scores");
+    for (unsigned int r = 0; r < n_r; ++r)
+      for (auto trial : _models[m]->getModelData<std::vector<std::vector<Real>>>("cv_scores"))
+        (*_cv_scores[n_r * m + r]).push_back(trial[_response_indices[r]]);
```
Comment (zachmprince) [inline modules/stochastic_tools/test/src/reporters/CrossValidationScores.C:49]:
```cpp
  for (unsigned int m = 0; m < _models.size(); ++m)
    (*_cv_scores[m]) = _models[m]->getModelData<std::vector<std::vector<Real>>>("cv_scores");
```

File: modules/stochastic_tools/src/surrogates/NearestPointTrainer.C [hunk 2]
```diff
@@ -17,70 +17,55 @@ NearestPointTrainer::validParams()
...
-  for (const auto & col : _predictor_cols)
-    _sample_points[d++][_local_row] = _sampler_row[col];
+  for (auto d : make_range(_n_dims))
+    _sample_points[d].push_back(_predictor_row[d]);
 
-  _sample_points.back()[_local_row] = _response;
+  // Get responses
+  if (_rval)
+    _sample_results[0].push_back(*_rval);
+
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/NearestPointTrainer.C:57]: Remove empty line

File: modules/stochastic_tools/src/surrogates/NearestPointTrainer.C [hunk 3]
```diff
@@ -17,70 +17,55 @@ NearestPointTrainer::validParams()
...
 void
 NearestPointTrainer::preTrain()
 {
-  // Resize to number of sample points
   for (auto & it : _sample_points)
-    it.resize(_sampler.getNumberOfLocalRows());
+    it.clear();
+
+  for (auto & it : _sample_results)
+    it.clear();
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/NearestPointTrainer.C:41]: Should we be doing a reserve here too?

File: modules/stochastic_tools/src/surrogates/NearestPointSurrogate.C [hunk 4]
```diff
@@ -22,19 +22,37 @@ NearestPointSurrogate::validParams()
...
+              "Input point does not match dimensionality of training data.");
+
+  y.assign(_sample_results.size(), 0.0);
+
+  for (const auto & r : make_range(y.size()))
+    y[r] = findNearestPoint(x, r);
+}
+
+Real
+NearestPointSurrogate::findNearestPoint(const std::vector<Real> & x, const unsigned int r) const
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/NearestPointSurrogate.C:53]: Let's make this slightly more efficient. Instead of returning the value, we should return the index in `_sample_points` then the `evaluate` functions can just use it to into `_sample_results`.

File: modules/stochastic_tools/examples/surrogates/cross_validation/gold/normal_train_and_cv_out_cv_scores_0002.csv [hunk 5]
```diff
@@ -1,4 +1,4 @@
-pr_surr
+pr_surr_0
```
Comment (zachmprince) [inline modules/stochastic_tools/examples/surrogates/cross_validation/gold/normal_train_and_cv_out_cv_scores_0002.csv:1]: Do we need these polynomial regression-only tests anymore?

**Review by zachmprince — CHANGES_REQUESTED** [review]
Very simple stuff here. Great work!

**Comment by crb3874** [comment]
Sorry this took so long to get addressed - I'm glad we decided to go through the analysis with all trainer types to add to the documentation, but it took a little while - the main limitation was that the configuration I'd used before (1e5 trials with 6560 sampler rows) was significantly more expensive to run with GP and ANN, and was probably overkill to begin with - I reduced the problem to 100 repetitions/1000 sampler rows to make it tractable. I only included results with uniform parameters because the results with normal parameters wasn't very interesting.

I also had issues with accuracy/convergence with the neural network. I got some help from Peter on setting better input parameters, and went ahead and implemented data standardization for ANN (not included in this PR) - this improved the results I was getting significantly. 

Otherwise, I believe I addressed all the requested changes. I removed the regression tests for PR-only, but left the input files in the examples folder since they're referred to in the documentation. I also re-golded as necessary to handle the change to JSON output for CrossValidationScores. Otherwise, I agree with all the changes/they're in this recent commit.

Thanks!

**Review by zachmprince — DISMISSED** [review]

**Review by zachmprince — APPROVED** [review]
